Threaded ASIC IRQ handlers

pcercuei opened the pull request:
Update `irq_set_handler`, `irq_set_global_handler`, `asic_evt_set_handler`, `vblank_handler_add` to accept an extra argument, that will be passed along to the registered callback.

This is very useful to pass state structures to the IRQ handlers, which can then pass it down the functions.

This will later be used to pass structures between a first-level IRQ handler and second-level threaded IRQ handlers, when those are implemented.

pcercuei commented:
Rebased and updated to use the ugly void-casts for unused parameters.

QuzarDC commented on kernel/arch/dreamcast/hardware/asic.c:
```diff
@@ -110,26 +113,46 @@
+    void *data;
+} asic_evt_handler_entry_t;
+
+struct asic_thdata {
```
Is there a compelling reason to do it this way? In this way it's not possible to determine whether or not a given handler is threaded or not (and if we need to call its own free/remove) or not. By expanding the `asic_evt_handler_entry_t` to have some component (maybe flags or a pointer for asic_thdata) the management of the handlers could be more easily be managed. 

I presume some sort of change to this with my comment at the end. But am open to having missed why this would be better or that's really unnecessary.

pcercuei replied:
My reason was keeping RAM usage low, but it's probably a better idea to add a field, yes.

QuzarDC replied:
Checking against the dispatch function being installed makes fine sense instead of an additional field. If it ever becomes an issue to need to know elsewhere, a function could be exposed to make that same check.

QuzarDC commented on kernel/arch/dreamcast/include/dc/asic.h:
```diff
@@ -193,9 +195,17 @@ typedef void (*asic_evt_handler)(uint32_t code);
 */
-void asic_evt_set_handler(uint16_t code, asic_evt_handler handler);
+void asic_evt_set_handler(uint16_t code, asic_evt_handler handler, void *data);
+
```
These two would need to get fleshed out dox.

pcercuei replied:
What do you mean? It's already documented.

pcercuei replied:
Nevermind - just noticed you were talking about the two functions below this one.

QuzarDC replied:
> Nevermind - just noticed you were talking about the two functions below this one.

Right, sorry, it made more sense in the review view where these are all inlined so the comment would have appeared directly in the spot where the dox go. I forgot they get truncated down in the normal views -_-;

QuzarDC reviewed (commented):
Looks great. Needs a bit more (the dox for the new functions). And I gave two other suggestions (unique thread names for the handlers + adjusting the data structure to detect type/collapse remove function).

pcercuei commented on kernel/arch/dreamcast/hardware/asic.c:
```diff
@@ -230,3 +252,99 @@ void asic_init(void) {
+    return 0;
+}
+
+void asic_remove_threaded_handler(uint16_t code)
```
Your `asic_remove_handler` wouldn't know which handler is a threaded interrupt and which one is not, because it only has a `void *data` pointer...

QuzarDC replied:
That was meant to be in the context of the other comment to add a way to distinguish them, but you figured that out by the end :) . My thinking was that for the handlers we want to make threaded the use cases we have seen have been that we want user-installable callbacks. It would make sense in those cases to use a regular handler by default ( which would likely be something like `{ ack_and_mask_func(); unmask_func(); }` ) then if a user requests a custom callback install the threaded version.

Overly long comment is just to document that reasoning for posterity.

pcercuei commented:
@QuzarDC I updated the PR with your suggestions and some more:
- It now sets the thread name and uses it in genwait_wait();
- Added documentation for the two new functions;
- I renamed `asic_remove_threaded_handler` to `asic_evt_remove_handler`;
- I renamed `asic_request_threaded_handler` to `asic_evt_request_threaded_handler`;
- `asic_evt_remove_handler` will work with both threaded and non-threaded handlers. I did not have to add a new field to `asic_evt_handler_entry_t`, because we know that threaded handlers will always have `.hnd = asic_thirq_dispatch`.
